fog/fog_types/src/view.rs

```diff
@@ -190,69 +193,178 @@ pub struct TxOutRecord {
-            e_fog_hint: EncryptedFogHint::from(&[0u8; ENCRYPTED_FOG_HINT_LEN]),
+impl FogTxOut {
+    /// Try to recover a TxOut from a FogTxOut and the user's private view key.
+    ///
+    /// * The amount commitment data is reconstructed, then we check if the
+    ///   reconstructed data matches the crc32 provided.
+    /// * The encrypted fog hint data is zeroed since it is not reconstructible
+    ///   and isn't needed by the client.
+    ///
+    /// Arguments:
+    /// * view_key: the private view key of the recipient of this TxOut
+    ///
+    /// Returns:
+    /// * TxOut,
+    /// Or
+    /// * An error if recovery failed
+    pub fn try_recover_tx_out(&self, view_key: &RistrettoPrivate) -> Result<TxOut, FogTxOutError> {
+        // Reconstruct compressed commitment based on our view key.
+        // The first step is reconstructing the TxOut shared secret
+        let public_key = RistrettoPublic::try_from(&self.public_key)?;
+        let tx_out_shared_secret =
+            mc_transaction_core::get_tx_out_shared_secret(view_key, &public_key);
+
+        // The next step is unblinding the amount value
+        let value =
+            self.amount_masked_value ^ mc_transaction_core::get_value_mask(&tx_out_shared_secret);
+
+        // Now we can rebuild the Amount object from the value and shared secret
+        let amount = Amount::new(value, &tx_out_shared_secret)?;
+
+        // Check that the crc32 of amount compressed commitment matches
+        if self.amount_commitment_data_crc32
+            != crc32::checksum_ieee(amount.commitment.point.as_bytes())
+        {
+            return Err(FogTxOutError::ChecksumMismatch);
         }
+
+        Ok(TxOut {
+            amount,
+            target_key: self.target_key,
+            public_key: self.public_key,
+            e_fog_hint: EncryptedFogHint::from(&[0u8; ENCRYPTED_FOG_HINT_LEN]),
+        })
+    }
+}
+
+/// An error that occurs when trying to convert a FogTxOut to a TxOut
+#[derive(Display, Debug)]
+pub enum FogTxOutError {
+    /// CompressedCommitment crc32 mismatch
+    ChecksumMismatch,
+    /// An invalid amount: {0}
+    Amount(AmountError),
+    /// An invalid key: {0}
+    Key(KeyError),
+}
+
+impl From<AmountError> for FogTxOutError {
+    fn from(src: AmountError) -> Self {
+        Self::Amount(src)
+    }
+}
+
+impl From<KeyError> for FogTxOutError {
+    fn from(src: KeyError) -> Self {
+        Self::Key(src)
     }
 }
+
+/// A collection of metadata about a TxOut that fog preserves in the TxOutRecord
+#[derive(Clone, Debug, Default, Eq, Hash, PartialEq)]
+pub struct FogTxOutMetadata {
+    /// The global index of this TxOut in the set of all TxOut's in the
+    /// blockchain
+    pub global_index: u64,
+    /// The index of the block in which this TxOut appeared
+    pub block_index: u64,
+    /// The timestamp of the block in which this TxOut appeared, in seconds
+    /// since the Unix epoch.
+    pub timestamp: u64,
+}
```

Comment:
this is added in order to have a nice constructor function for TxOutRecord, otherwise you take three different u64's as arguments and if they are in the wrong order, you have a bug